component/storage/leveldb/leveldb.go

```diff
@@ -0,0 +1,503 @@
+	}
+
+	return nil
+}
+
+// GetStoreConfig returns the current store configuration.
+func (p *Provider) GetStoreConfig(name string) (storage.StoreConfiguration, error) {
+	name = strings.ToLower(name)
+
+	openStore, ok := p.dbs[name]
+	if !ok {
+		return storage.StoreConfiguration{}, storage.ErrStoreNotFound
+	}
+
+	storeConfigBytes, err := openStore.Get(storeConfigKey)
+	if err != nil {
+		return storage.StoreConfiguration{},
+			fmt.Errorf(`failed to get store configuration for "%s": %w`, name, err)
+	}
+
+	var storeConfig storage.StoreConfiguration
+
+	err = json.Unmarshal(storeConfigBytes, &storeConfig)
+	if err != nil {
+		return storage.StoreConfiguration{}, fmt.Errorf("failed to unmarshal store configuration: %w", err)
+	}
+
+	return storeConfig, nil
+}
+
+// GetOpenStores is not implemented.
+func (p *Provider) GetOpenStores() []storage.Store {
+	panic("implement me")
+}
+
+// Close closes all stores created under this store provider.
+func (p *Provider) Close() error {
+	p.lock.RLock()
+
+	openStoresSnapshot := make([]*store, len(p.dbs))
+
+	var counter int
+
+	for _, openStore := range p.dbs {
+		openStoresSnapshot[counter] = openStore
+		counter++
+	}
+	p.lock.RUnlock()
+
+	for _, openStore := range openStoresSnapshot {
+		err := openStore.Close()
+		if err != nil {
+			return fmt.Errorf(`failed to close open store with name "%s": %w`, openStore.name, err)
+		}
+	}
+
+	return nil
+}
+
+// getLeveldbStore finds level db store with given name
+// returns nil if not found.
+func (p *Provider) getLeveldbStore(name string) *store {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	return p.dbs[name]
+}
+
+// newLeveldbStore creates level db store for given name space
+// returns nil if not found.
+func (p *Provider) newLeveldbStore(name string) (*store, error) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	db, err := leveldb.OpenFile(fmt.Sprintf(pathPattern, p.dbPath, name), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	store := &store{db: db, name: name, close: p.removeStore}
+	p.dbs[name] = store
+
+	// Create the tag map if it doesn't exist already.
+	_, err = store.Get(tagMapKey)
+	if errors.Is(err, storage.ErrDataNotFound) {
+		err = store.Put(tagMapKey, []byte("{}"))
+		if err != nil {
+			return nil, fmt.Errorf(`failed to create tag map for "%s": %w`, name, err)
+		}
+	} else if err != nil {
+		return nil, fmt.Errorf("unexpected failure while getting tag data bytes: %w", err)
+	}
+
+	return store, nil
+}
+
+func (p *Provider) removeStore(name string) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	_, ok := p.dbs[name]
+	if ok {
+		delete(p.dbs, name)
+	}
+}
+
+type store struct {
+	db    *leveldb.DB
+	name  string
+	close closer
+}
+
+// Put stores the key and the record.
+func (s *store) Put(key string, value []byte, tags ...storage.Tag) error {
+	if key == "" {
+		return errors.New("key cannot be blank")
+	}
+
+	if value == nil {
+		return errors.New("value cannot be nil")
+	}
```

Comment:
Why value cannot be blank/nil? Should be removed.

Reply:
There's a couple reasons...
1. For the Batch method defined in the interface, a `nil` value is special and indicates a deletion (See https://github.com/hyperledger/aries-framework-go/blob/9009c86eef13eaee1b7f7d686ba74f4c4f68f20c/spi/storage/storage.go#L66). So a "nil" value has special meaning. If you were to replace a `Put` call with a `Batch` call, the behaviour should be the same. Imagine if you were storing a nil value before, and then suddenly it becomes a deletion... So for consistency, I think it's best if you can't store a `nil` value. I already defined in the interface that this is an error (https://github.com/hyperledger/aries-framework-go/blob/9009c86eef13eaee1b7f7d686ba74f4c4f68f20c/spi/storage/storage.go#L100) before as well.
2. I have common storage tests ensuring that a nil value causes an error already (https://github.com/hyperledger/aries-framework-go/blob/9009c86eef13eaee1b7f7d686ba74f4c4f68f20c/test/component/storage/storage.go#L407), and those are used here, so to keep them passing I have to maintain it.